proposals/0698-qualified-literals.rst

```diff
@@ -0,0 +1,359 @@
+    :align: left
+
+    * - **New expression syntax**
+      - **Desugared expression syntax**
+    * - ``Foo.1``
+      - ``Foo.fromNatural 1``
+    * - ``Foo.(1)``
+      - ``Foo.fromNatural (1)``
+    * - ``Foo.(-1)``
+      - ``Foo.fromNegativeInteger (-1)``
+    * - ``Foo.(1.2)``
+      - ``Foo.fromRational (1.2)``
+    * - ``Foo."asdf"``
+      - ``Foo.fromString "asdf"``
+    * - ``Foo."""asdf"""``
+      - ``Foo.fromString "asdf"``
+    * - ``Foo.[x, y]``
+      - ``Foo.fromList (x `Foo.listCons` (y `Foo.listCons` Foo.listNil))``
+
+And the following syntaxes for patterns
+
+.. list-table::
+    :align: left
+
+    * - **New pattern syntax**
+      - **Desugared ViewPattern**
+    * - ``Foo.1``
+      - ``(Foo.toNatural -> 1)``
+    * - ``Foo.(1)``
+      - ``(Foo.toNatural -> (1))``
+    * - ``Foo.(-1)``
+      - ``(Foo.toNegativeInteger -> (-1))``
+    * - ``Foo.(1.2)``
+      - ``(Foo.toRational -> (1.2))``
+    * - ``Foo."asdf"``
+      - ``(Foo.toString -> "asdf")``
+    * - ``Foo."""asdf"""``
+      - ``(Foo.toString -> "asdf")``
+    * - ``Foo.[x, y]``
+      - ``(Foo.listUncons -> Just (x, Foo.listUncons -> Just (y, Foo.listUncons -> Nothing)))``
+    * - ``Foo.(x : _)``
+      - ``(Foo.listUncons -> Just (x, _))``
+
+As long as the desugared expressions/patterns type check, users are free to define these functions however they want.
+
+Notes:
+
+* Parentheses are required for negative integers and rationals, to avoid ambiguity, both in the lexer and for human readers. Parentheses are optional for positive integers.
+
+* Multiline strings are desugared to single line strings first, then desugared as a qualified string literal.
+
+* Some literals are not supported yet (Chars, unboxed literals) due to lack of use-cases, but could be extended in the future.
+
+* Future work could be done to allow compile time logic, e.g. ``$Foo.1`` => ``$(Foo.fromInteger [|1|])``, but that is out of scope of this proposal.
+
+Parser
+~~~~~~
+
+Update `Section 10.5 <https://www.haskell.org/onlinereport/haskell2010/haskellch10.html#x17-18000010.5>`_ of the Haskell 2010 report as follows.
+
+.. code-block:: abnf
+
+  aexp → qvar
+       | ...
+       | modid . integer
+       | modid . ( {-} integer )
+       | modid . ( float )
+       | modid . string
+       | modid . multiLineString
+       | modid . [ exp_1 , ..., exp_k ]
+
+  apat → var [ @ apat ]
+       | ...
+       | modid . integer
+       | modid . ( {-} integer )
+       | modid . ( float )
+       | modid . string
+       | modid . multiLineString
+       | modid . [ pat_1 , ..., pat_k ]
+       | modid . ( pat_1 : ... : pat_k )
+
+
+Proposed Library Change Specification
+-------------------------------------
+
+Template Haskell
+~~~~~~~~~~~~~~~~
+
+The following constructors will be updated to the following definitions:
+
+::
+
+  ListE (Maybe ModuleName) [Exp]
+
+  StringL (Maybe ModuleName) String
+  IntegerL (Maybe ModuleName) Integer
+  RationalL (Maybe ModuleName) Rational
```

Comment:
Rather than changing these constructors, could we add new ones, to avoid a breaking change for existing clients?